src/main/java/org/breedinginsight/services/impl/BreedingMethodServiceImpl.java

```diff
@@ -60,28 +60,39 @@ public List<ProgramBreedingMethodEntity> fetchBreedingMethodsInUse(UUID programI
     }
 
     @Override
-    public ProgramBreedingMethodEntity createBreedingMethod(ProgramBreedingMethodEntity breedingMethod, UUID programId, UUID userId) throws BadRequestException, ApiException {
-        if(!validateBreedingMethod(breedingMethod)) {
-            throw new BadRequestException("Missing required data");
-        }
+    public ProgramBreedingMethodEntity createBreedingMethod(ProgramBreedingMethodEntity breedingMethod, UUID programId, UUID userId) throws BadRequestException {
+        validate(breedingMethod, programId);
 
         return dsl.transactionResult(() -> breedingMethodDAO.createProgramMethod(breedingMethod, programId, userId));
     }
 
+
     @Override
     public ProgramBreedingMethodEntity updateBreedingMethod(ProgramBreedingMethodEntity breedingMethod, UUID programId, UUID userId) throws BadRequestException, ApiException {
-        if(!validateBreedingMethod(breedingMethod)) {
-            throw new BadRequestException("Missing required data");
-        }
-
         List<ProgramBreedingMethodEntity> inUseMethods = fetchBreedingMethodsInUse(programId);
         if(inUseMethods.stream().anyMatch(method -> method.getId().equals(breedingMethod.getId()))) {
             throw new BadRequestException("Breeding method is not allowed to be edited");
         }
+        validate(breedingMethod, programId);
 
         return dsl.transactionResult(() -> breedingMethodDAO.updateProgramMethod(breedingMethod, programId, userId));
     }
 
+    private void validate(ProgramBreedingMethodEntity breedingMethod, UUID programId) throws BadRequestException {
+        if (isMissingRequiredFields(breedingMethod)) {
+            throw new BadRequestException("Missing required data");
+        }
+
+        List<ProgramBreedingMethodEntity> programAndSystemMethods = getBreedingMethods(programId);
+        if( isDuplicateMethodNameFoundOnList(breedingMethod, programAndSystemMethods)){
+            throw new BadRequestException(format("'%s' is already defined in the system.", breedingMethod.getName()));
+        }
+        if( isDuplicateMethodAbbreviationFoundOnList(breedingMethod, programAndSystemMethods)){
+            throw new BadRequestException(format("'%s' is already defined in the system.", breedingMethod.getAbbreviation()));
+        }
```

Comment:
I think it would be nice to have more descriptive error messages. If a user enters a duplicate *name*, tell them that the *name* is a duplicate, and same for abbreviation. Maybe ask Shawn to add the desired language to the source of truth documentation, I didn't see these error messages documented anywhere.

Reply:
Thanks.

The messages are now:
`A breeding method with the abbreviation '<abbreviation>' is already defined in the system.`
`A breeding method with the name '<name>' is already defined in the system.`